File: 백준/Gold/4485. 녹색 옷 입은 애가 젤다지？/녹색 옷 입은 애가 젤다지？.py
import sys
input=sys.stdin.readline
from heapq import heappop, heappush

def dijkstra(g, n):
    q=[]
    vis=[[INF]*n for _ in range(n)]
    vis[0][0]=g[0][0]
    heappush(q, (g[0][0], [0, 0]))
    while q:
        distance, node = heappop(q)
        if vis[node[0]][node[1]]<distance:
            continue
        for i in range(4):
            nx, ny = node[0]+dx[i], node[1]+dy[i]
            if 0<=nx<n and 0<=ny<n:
                cost=distance+g[nx][ny]
                if cost<vis[nx][ny]:
                    vis[nx][ny]=cost
                    heappush(q, (cost, [nx, ny]))
    return vis

INF=int(1e9)
dx=[1, -1, 0, 0]
dy=[0, 0, 1, -1]
i=1
while True:
    n=int(input())
    if n==0:
        break
    g=[list(map(int, input().split())) for _ in range(n)]
    print(f'Problem {i}:',dijkstra(g, n)[-1][-1])
    i+=1

File: 백준/Gold/4485. 녹색 옷 입은 애가 젤다지？/녹색 옷 입은 애가 젤다지？.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.PriorityQueue;
import java.util.StringTokenizer;

public class Main {
	static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
	static StringBuilder sb = new StringBuilder();
	static StringTokenizer st;
	static int n, g[][];
	static int[] dx = { 0, 1, 0, -1 }, dy = { 1, 0, -1, 0 };

	public static void main(String[] args) throws Exception {
		int idx = 0;
		while (true) {
			idx++;
			n = Integer.parseInt(br.readLine());
			if (n == 0) {
				break;
			}
			g = new int[n][n];
			for (int i = 0; i < n; i++) {
				st = new StringTokenizer(br.readLine());
				for (int j = 0; j < n; j++) {
					g[i][j] = Integer.parseInt(st.nextToken());
				}
			}
			sb.append("Problem " + idx + ": " + dijkstra()).append("\n");
		}
		System.out.println(sb);
	}

	static int dijkstra() {
		int[][] vis = new int[n][n];
		for (int i = 0; i < n; i++) {
			for (int j = 0; j < n; j++) {
				vis[i][j] = -1;
			}
		}
		PriorityQueue<Node> pq = new PriorityQueue<Node>();
		pq.add(new Node(new int[] { 0, 0 }, g[0][0]));
		vis[0][0] = g[0][0];
		while (!pq.isEmpty()) {
			Node node = pq.poll();
			int xx = node.to[0], yy = node.to[1], dd = node.dist;
			for (int i = 0; i < 4; i++) {
				int nx = xx + dx[i], ny = yy + dy[i];
				if (0 <= nx && nx < n && 0 <= ny && ny < n) {
					int cost = dd + g[nx][ny];
					if (vis[nx][ny] == -1 || cost < vis[nx][ny]) {
						pq.add(new Node(new int[] { nx, ny }, cost));
						vis[nx][ny] = cost;
					}
				}
			}
		}
		return vis[n - 1][n - 1];
	}
}

class Node implements Comparable<Node> {
	int[] to;
	int dist;

	public Node(int[] n, int d) {
		this.to = n;
		this.dist = d;
	}

	public int compareTo(Node o) {
		return Integer.compare(this.dist, o.dist);
	}
}